sign/cosi/cosi.go

```diff
@@ -1,452 +1,382 @@
+// of public keys and cosigning policy.
+func Verify(suite Suite, publics []kyber.Point, message, sig []byte, policy Policy) error {
+	if publics == nil {
+		return errors.New("no public keys provided")
+	}
+	if message == nil {
+		return errors.New("no message provided")
+	}
+	if sig == nil {
+		return errors.New("no signature provided")
+	}
+	if policy == nil {
+		policy = CompletePolicy{}
+	}
 
-	if !left.Equal(c.aggregateCommitment) {
-		return errors.New("recreated commitment is not equal to one given")
+	lenCom := suite.PointLen()
+	VBuff := sig[:lenCom]
+	V := suite.Point()
+	if err := V.UnmarshalBinary(VBuff); err != nil {
+		return errors.New("unmarshalling of commitment failed")
 	}
 
-	return nil
-}
+	// Unpack the aggregate response
+	lenRes := lenCom + suite.ScalarLen()
+	rBuff := sig[lenCom:lenRes]
+	r := suite.Scalar().SetBytes(rBuff)
 
-// VerifySignature is the method to call to verify a signature issued by a Cosi
-// struct. Publics is the WHOLE list of publics keys, the mask at the end of the
-// signature will take care of removing the indivual public keys that did not
-// participate
-func VerifySignature(group kyber.Group, publics []kyber.Point, message, sig []byte) error {
-	aggCommitBuff := sig[:32]
-	aggCommit := group.Point()
-	if err := aggCommit.UnmarshalBinary(aggCommitBuff); err != nil {
-		panic(err)
-	}
-	sigBuff := sig[32:64]
-	sigInt := group.Scalar().SetBytes(sigBuff)
-	maskBuff := sig[64:]
-	mask := newMask(group, publics)
-	mask.SetMask(maskBuff)
-	aggPublic := mask.Aggregate()
-	aggPublicMarshal, err := aggPublic.MarshalBinary()
+	// Unpack the participation mask and get the aggregate public key
+	mask, err := NewMask(suite, publics, nil)
 	if err != nil {
 		return err
 	}
+	mask.SetMask(sig[lenRes:])
+	A := mask.AggregatePublic
+	ABuff, err := A.MarshalBinary()
+	if err != nil {
+		return errors.New("marshalling of aggregate public key failed")
+	}
 
-	hash := sha512.New()
-	hash.Write(aggCommitBuff)
-	hash.Write(aggPublicMarshal)
+	// Recompute the challenge
+	hash := suite.Hash()
+	hash.Write(VBuff)
+	hash.Write(ABuff)
 	hash.Write(message)
 	buff := hash.Sum(nil)
-	k := group.Scalar().SetBytes(buff)
+	k := suite.Scalar().SetBytes(buff)
 
 	// k * -aggPublic + s * B = k*-A + s*B
 	// from s = k * a + r => s * B = k * a * B + r * B <=> s*B = k*A + r*B
 	// <=> s*B + k*-A = r*B
-	minusPublic := group.Point().Neg(aggPublic)
-	kA := group.Point().Mul(k, minusPublic)
-	sB := group.Point().Mul(sigInt, nil)
-	left := group.Point().Add(kA, sB)
+	minusPublic := suite.Point().Neg(A)
+	kA := suite.Point().Mul(k, minusPublic)
+	sB := suite.Point().Mul(r, nil)
+	left := suite.Point().Add(kA, sB)
 
-	if !left.Equal(aggCommit) {
-		return errors.New("Signature invalid")
+	x, err := left.MarshalBinary()
+	if err != nil {
+		return err
+	}
+	y, err := V.MarshalBinary()
+	if err != nil {
+		return err
+	}
+	if subtle.ConstantTimeCompare(x, y) == 0 || !policy.Check(mask) {
```

Comment:
why subtle here ? I don't see any benefits because it does not deal with any secret inputs, and costs you 2 `MarshalBinary` and a slow compare.